presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTimeWithTimeZone.java

```diff
@@ -119,18 +113,36 @@ public boolean equals(Object obj)
     @Override
     public String toString()
     {
+        Instant utcInstantInCurrentDay = getTimeInstantInCurrentDay();
+        ZonedDateTime zonedDateTime = utcInstantInCurrentDay.atZone(ZoneId.of(timeZoneKey.getId()));
+        return zonedDateTime.format(formatter);
+    }
+
+    private Instant getTimeInstantInCurrentDay()
+    {
         // FIXME This is hack that we need to use as the timezone interpretation depends on date (not only on time)
         // Additionally we cannot just grab current TZ offset from timezoneOffsetReferenceTimestampUtc and use it
         // as our formatting library will fail to display expected TZ symbol.
         long currentMillisOfDay =
-                LocalTime.from(
+                TimeUnit.NANOSECONDS.toMillis(LocalTime.from(
                         Instant.ofEpochMilli(timezoneOffsetReferenceTimestampUtc)
                                .atZone(ZoneOffset.UTC))
-                        .toNanoOfDay() / 1_000_000L; // nanos to millis
+                        .toNanoOfDay());
```

Comment:
Just  a reminder: i still think that this can be written simply as
```
currentMillisOfDay = ChronoField.MILLI_OF_DAY.getFrom(
    Instant.ofEpochMilli(timezoneOffsetReferenceTimestampUtc).atZone(ZoneOffset.UTC))
```